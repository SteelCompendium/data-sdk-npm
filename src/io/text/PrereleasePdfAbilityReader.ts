import { Ability } from "../../model/Ability";
import { IDataReader } from "../IDataReader";
import { Effects } from "../../model/Effects";
import { Effect, MundaneEffect, PowerRollEffect } from "../../model";

export class PrereleasePdfAbilityReader implements IDataReader<Ability> {
    read(text: string): Ability {
        const lines = text.split('\n').map(l => l.trim()).filter(l => l);
        const effects: Effect[] = [];
        const abilityData: Partial<Ability> = {};

        if (!lines.length) return new Ability({ ...abilityData, effects: new Effects([]) });

        const firstLine = lines.shift() || '';
        const costMatch = firstLine.match(/\((.*)\)/);
        if (costMatch) {
            abilityData.cost = costMatch[1].trim();
            abilityData.name = firstLine.replace(costMatch[0], '').trim();
        } else {
            abilityData.name = firstLine.trim();
        }

        let inEffectSection = false;
        const propertyLines: string[] = [];
        const effectLines: string[] = [];

        for (const line of lines) {
            if (line.toLowerCase().startsWith('effect:')) {
                inEffectSection = true;
                const effectText = line.substring('effect:'.length).trim();
                if (effectText) effectLines.push(line);
                continue;
            }
            if (line.startsWith('•')) {
                inEffectSection = true;
            } else {
                const namedEffectMatch = line.match(/^([a-zA-Z0-9\s]+):\s(.+)/);
                const powerRollTierMatch = line.match(/^((\d+-\d+)|(\d+ or lower)|(\d+ or higher)|(\d+\+)|crit|critical)/i)
                if (namedEffectMatch && !powerRollTierMatch) {
                    inEffectSection = true;
                }
            }

            if (inEffectSection) {
                effectLines.push(line);
            } else {
                propertyLines.push(line);
            }
        }

        const types = ['Action', 'Maneuver', 'Triggered Action', 'Villain Action 1', 'Main Action'];
        const flavor: string[] = [];
        for (const line of propertyLines) {
            let isProperty = false;

            const distanceMatch = line.match(/Distance:\s*(.*?)(?:\s*Target:|$)/i);
            if (distanceMatch && distanceMatch[1]) {
                abilityData.distance = distanceMatch[1].trim();
                isProperty = true;
            }

            const targetMatch = line.match(/Target:\s*(.*)/i);
            if (targetMatch && targetMatch[1]) {
                abilityData.target = targetMatch[1].trim();
                isProperty = true;
            }

            const keywords: string[] = [];
            const foundTypes: string[] = [];
            const parts = line.split(',').map(p => p.trim());
            const likelyKeywordLine = parts.length > 1 && parts.every(p => p.split(' ').length < 5);

            if (likelyKeywordLine) {
                parts.forEach(part => {
                    let typeFound = false;
                    for (const type of types) {
                        if (part.toLowerCase().includes(type.toLowerCase())) {
                            foundTypes.push(type === 'Main Action' ? 'Action' : type);
                            const remaining = part.replace(new RegExp(type, 'i'), '').trim();
                            if (remaining) keywords.push(remaining);
                            typeFound = true;
                            isProperty = true;
                            break;
                        }
                    }
                    if (!typeFound) keywords.push(part);
                });
            }

            if (foundTypes.length > 0) {
                abilityData.type = foundTypes.join(', ');
                abilityData.keywords = (abilityData.keywords || []).concat(keywords.filter(k => k));
            } else if (likelyKeywordLine) {
                abilityData.keywords = (abilityData.keywords || []).concat(keywords.filter(k => k));
                isProperty = true;
            }

            if (!isProperty) {
                flavor.push(line);
            }
        }

        if (flavor.length > 0) abilityData.flavor = flavor.join(' ');

        if (effectLines.length > 0) {
            let currentEffect: { name?: string, effect: string, roll?: string } & Record<string, string> | null = null;
            let currentEffectLines: string[] = [];

            const flushEffect = () => {
                if (currentEffectLines.length > 0) {
                    const firstLine = currentEffectLines[0];

                    const powerRollMatch = firstLine.match(/Power Roll\s*\+\s*([a-zA-Z]+)/i);
                    if (powerRollMatch) {
                        currentEffect = { roll: `Power Roll + ${powerRollMatch[1]}` } as any;
                        for (let i = 1; i < currentEffectLines.length; i++) {
                            const line = currentEffectLines[i];
                            const tierMatch = line.match(/^((?:\d+-\d+)|(?:\d+ or lower)|(?:\d+\+)|crit(?:ical)?):\s*(.*)/i);
                            if (tierMatch && currentEffect) {
                                const key = tierMatch[1].replace(' or lower', ' or lower').replace('critical', 'crit');
                                currentEffect[key] = tierMatch[2];
                            }
                        }
                        if (currentEffect) {
                            effects.push(new PowerRollEffect(currentEffect));
                        }
                    } else {
                        effects.push(new MundaneEffect({ effect: currentEffectLines.join(' ') }));
                    }
                }
                currentEffectLines = [];
            }


            for (const line of effectLines) {
                if (line.toLowerCase().startsWith('effect:')) {
                    flushEffect();
                    currentEffectLines.push(line.substring('effect:'.length).trim());
                } else if (line.startsWith('•')) {
                    flushEffect();
                    currentEffectLines.push(line.substring(1).trim());
                } else {
                    const namedMatch = line.match(/^([a-zA-Z0-9\s]+):\s(.+)/);
                    if (namedMatch) {
                        flushEffect();
                        currentEffectLines.push(line);
                    } else if (currentEffectLines.length > 0) {
                        currentEffectLines[currentEffectLines.length - 1] += ' ' + line;
                    }
                }
            }
            flushEffect();

        } else if (propertyLines.length > 0 && effects.length === 0) {
            effects.push(new MundaneEffect({ effect: propertyLines.join(' ') }));
            if (abilityData.flavor) delete abilityData.flavor;
            if (abilityData.keywords) delete abilityData.keywords;
            if (abilityData.distance) delete abilityData.distance;
            if (abilityData.target) delete abilityData.target;
        }

        if (!abilityData.type) {
            abilityData.type = 'Unknown';
        }

        const finalAbility = new Ability(abilityData);
        finalAbility.effects = new Effects(effects);
        return finalAbility;
    }
} 