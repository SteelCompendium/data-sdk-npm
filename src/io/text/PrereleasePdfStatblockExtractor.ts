import { Statblock } from "../../model";
import { IDataExtractor } from "../IDataExtractor";
import { PrereleasePdfStatblockReader } from "./PrereleasePdfStatblockReader";

export class PrereleasePdfStatblockExtractor implements IDataExtractor<Statblock> {
    extract(text: string): Statblock[] {
        const statblocks = PrereleasePdfStatblockExtractor.extractStatblockText(text);
        const reader = new PrereleasePdfStatblockReader();
        return statblocks.map(reader.read);
    }

    static extractStatblockText(text: string): string[] {
        const lines = text.split(/\r?\n/);

        const statblockStarts: { name: string; index: number }[] = [];
        const statblockHeaderRegex = /level\s+\d+\s+/i;
        const reader = new PrereleasePdfStatblockReader();

        lines.forEach((line, index) => {
            if (statblockHeaderRegex.test(line)) {
                // To be a statblock header, it must also have a name
                const lineWithNext = line + "\n" + (lines[index + 1] ?? "");
                try {
                    const statblock = reader.read(lineWithNext);
                    if (statblock.name && statblock.roles.length > 0) {
                        const existing = statblockStarts.find(s => s.name === statblock.name);
                        if (!existing) {
                            statblockStarts.push({ name: statblock.name, index });
                        }
                    }
                } catch (e) {
                    // Not a valid statblock header, ignore
                }
            }
        });

        return statblockStarts.map(({ name, index }, i) => {
            const start = index;
            const end = i < statblockStarts.length - 1 ? statblockStarts[i + 1].index : lines.length;
            const block = lines.slice(start, end).join("\n");
            return block;
        });
    }
}