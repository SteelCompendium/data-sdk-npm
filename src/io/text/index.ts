export * from './PrereleasePdfStatblockReader';
export * from './PrereleasePdfAbilityReader';