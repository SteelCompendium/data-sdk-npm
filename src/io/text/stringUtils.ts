export function cleanOcrText(text: string): string {
    let cleanedText = text;

    // 1. Fallback for mixed-case text to remove spaces after single consonants, avoiding articles like 'a'.
    // e.g. "B lade" -> "Blade".
    cleanedText = cleanedText.replace(/\b([B-HJ-NP-Z])\s+/g, '$1');

    // 2. Remove spaces before apostrophes. e.g. "don 't" -> "don't".
    cleanedText = cleanedText.replace(/\s+(?=[â€™'])/g, "");

    // 3. Remove spaces after a hyphenated prefix. e.g. "self- inflicted" -> "self-inflicted".
    cleanedText = cleanedText.replace(/(-[a-zA-Z])\s+/g, '$1');

    return cleanedText;
} 