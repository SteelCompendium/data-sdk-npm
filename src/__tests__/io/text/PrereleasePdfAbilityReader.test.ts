import { describe, it, expect } from "@jest/globals";
import fs from "fs";
import path from "path";
import { PrereleasePdfAbilityReader } from "../../../io/text/PrereleasePdfAbilityReader";

describe("PrereleasePdfAbilityReader", () => {
    const reader = new PrereleasePdfAbilityReader();
    const inputsDir = path.join(__dirname, "..", "..", "data", "ability", "prerelease-pdf");

    const testFiles = fs.readdirSync(inputsDir).filter(file => file.endsWith(".txt"));

    testFiles.forEach(file => {
        it(`should correctly parse ${file}`, () => {
            const inputPath = path.join(inputsDir, file);
            const inputText = fs.readFileSync(inputPath, "utf-8");
            const result = reader.read(inputText).toDTO();
            expect(result).toMatchSnapshot();
        });
    });
}); 