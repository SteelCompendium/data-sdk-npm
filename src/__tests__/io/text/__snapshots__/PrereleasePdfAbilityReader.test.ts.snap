// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PrereleasePdfAbilityReader should correctly parse distance_is_only_memory.txt 1`] = `
{
  "effects": [
    {
      "effect": "Each time you finish a respite, you can open a two-way portal that leads to any place you have previously been. You and your allies can pass through the portal, which remains open for 1 hour or until you dismiss it as a main action.",
    },
  ],
  "name": "DISTANCE IS ONLY MEMORY",
  "type": "Unknown",
}
`;

exports[`PrereleasePdfAbilityReader should correctly parse instantaneous_excavation.txt 1`] = `
{
  "cost": "5 ESSENCE",
  "effects": [
    {
      "effect": "Distance: Ranged 10 Target: Special",
    },
    {
      "roll": "Power Roll + Reason",
    },
    {
      "effect": "11 or lower: The target can shift 1 square from the edge of the hole to the nearest unoccupied space of their choice.",
    },
    {
      "effect": "12–16: The target falls into the hole.",
    },
    {
      "effect": "17+: The target falls into the hole and can’t reduce the height of the fall.",
    },
    {
      "effect": "Persistent 1: At the start of your turn, you open another hole, making a power roll against each creature who could fall into the hole when it opens without spending essence.",
    },
  ],
  "flavor": "The surface of the world around you opens up to swallow foes.",
  "keywords": [
    "Earth",
    "Magic",
    "Ranged",
  ],
  "name": "INSTANTANEOUS EXCAVATION",
  "type": "Maneuver",
}
`;

exports[`PrereleasePdfAbilityReader should correctly parse o_flower_aid_o_earth_defend.txt 1`] = `
{
  "cost": "5 ESSENCE",
  "effects": [
    {
      "effect": "Distance: 3 cube within 10 Target: Special",
    },
    {
      "effect": "Until the start of your next turn, the area gains the following effects:",
    },
    {
      "effect": "Once at the start of your turn as a free maneuver, you and each ally in the area can spend any number of Recoveries. Once as a free maneuver at the start of your turn, you allow yourself and each ally in the area to spend any number of Recoveries.",
    },
    {
      "effect": "The area is difficult terrain for enemies.",
    },
    {
      "effect": "Each enemy who enters the area for the first time in a combat round or starts their turn there takes damage equal to your Reason score.",
    },
    {
      "effect": "Persistent 1: The area remains until the start of your next turn. As a maneuver, you can move the area up to 5 squares. This ability ends if the area is ever not within your line of effect.",
    },
  ],
  "flavor": "Revitalizing plants and jagged stones grow, helping allies and hindering foes.",
  "keywords": [
    "Area",
    "Earth",
    "Green",
    "Magic",
    "Ranged",
  ],
  "name": "O FLOWER AID, O EARTH DEFEND",
  "type": "Maneuver",
}
`;

exports[`PrereleasePdfAbilityReader should correctly parse ray_of_agonizing_self_reflection.txt 1`] = `
{
  "effects": [
    {
      "roll": "Power Roll + Reason",
    },
    {
      "effect": "11 or lower: 2 + R corruption damage; R < weak, slowed (save ends)",
    },
    {
      "effect": "12–16: 4 + R corruption damage; R < average, slowed (save ends)",
    },
    {
      "effect": "17+: 6 + R corruption damage; R < strong, slowed (save ends)",
    },
  ],
  "flavor": "You inflict pain and doubt in equal measure.",
  "keywords": [
    "Magic",
    "Ranged",
    "Strike",
    "Void Main",
  ],
  "name": "RAY OF AGONIZING SELF-R EFLECTION",
  "type": "Action",
}
`;

exports[`PrereleasePdfAbilityReader should correctly parse there_is_no_space_between.txt 1`] = `
{
  "effects": [
    {
      "effect": "Distance: Ranged 10 Target: Special",
    },
    {
      "effect": "You open two size 1 portals in unoccupied spaces within distance, which last until you move beyond distance from any portal, end the effect as a maneuver, or are dying. Each portal must be placed at a height of no more than 1 square above the ground. When you or any ally touch a portal, that creature can choose to be instantly teleported to an unoccupied space of their choice adjacent to the other portal. If an enemy is force moved into a portal, their forced movement ends and they emerge from the other portal in an unoccupied space chosen by the creature who force moved them. At the start of each of your turns while the portals are active, you can open a new portal connected to the others. If three or more portals are present, you and your allies choose which portal to emerge from when entering a portal, and a creature who force moves an enemy into a portal chooses that enemy’s destination portal.",
    },
  ],
  "flavor": "Knowledge of the mystery reveals that two spaces are the same space.",
  "keywords": [
    "Magic",
    "Ranged",
    "Void",
  ],
  "name": "THERE IS NO SPACE BETWEEN",
  "type": "Maneuver",
}
`;

exports[`PrereleasePdfAbilityReader should correctly parse translated_through_flame.txt 1`] = `
{
  "cost": "5 ESSENCE",
  "effects": [
    {
      "effect": "Distance: Ranged 10 Target: Self or one ally",
    },
    {
      "roll": "Power Roll + Reason",
    },
    {
      "effect": "11 or lower: 3 fire damage",
    },
    {
      "effect": "12–16: 5 fire damage",
    },
    {
      "effect": "17+: 8 fire damage",
    },
  ],
  "flavor": "Your ally disappears, then reappears in a burst of fire.",
  "keywords": [
    "Fire",
    "Magic",
    "Ranged",
    "Void Main",
  ],
  "name": "TRANSLATED THROUGH FLAME",
  "type": "Action",
}
`;

exports[`PrereleasePdfAbilityReader should correctly parse unquiet_ground.txt 1`] = `
{
  "effects": [
    {
      "roll": "Power Roll + Reason",
    },
    {
      "effect": "11 or lower: 2 damage",
    },
    {
      "effect": "12–16: 5 damage",
    },
    {
      "effect": "17+: 7 damage",
    },
    {
      "effect": "The ground beneath the area is difficult terrain for enemies.",
    },
  ],
  "flavor": "A sudden storm of detritus assaults your foes and leaves them struggling to move.",
  "keywords": [
    "Area",
    "Earth",
    "Magic",
    "Ranged Main",
  ],
  "name": "UNQUIET GROUND",
  "type": "Action",
}
`;

exports[`PrereleasePdfAbilityReader should correctly parse viscious_fire.txt 1`] = `
{
  "effects": [
    {
      "roll": "Power Roll + Reason",
    },
    {
      "effect": "11 or lower: 2 + R fire damage; push 2",
    },
    {
      "effect": "12–16: 5 + R fire damage; push 3",
    },
    {
      "effect": "17+: 7 + R fire damage; push 4",
    },
  ],
  "flavor": "A jet of heavy fire erupts where you strike.",
  "keywords": [
    "Fire",
    "Magic",
    "Ranged",
    "Strike Main",
  ],
  "name": "VISCOUS FIRE",
  "type": "Action",
}
`;
